### PR TITLE
stellar-core: 0.5.1 -> 17.0.0

### PR DESCRIPTION
**Pull request by austinbutler**
###### Motivation for this change

Upgrades to latest and also fixes the build.

ZHF: #122042

@NixOS/nixos-release-managers 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by austinbutler** [comment]
Result of `nixpkgs-review pr 123294` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>stellar-core</li>
  </ul>
</details>

**Comment by austinbutler** [comment]
While this builds on my Arch machine with Nix installed, it fails on my NixOS machine with a wrong sha256 sum which is strange... Currently trying to troubleshoot that.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 123294` at d44c5319 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/a1336fd05080ca7fcfb508297b2d493c">stellar-core</a></li>
  </ul>
</details>
<details>
  <summary>5 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

autoconf is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/blockchains/stellar-core.nix:20:3:
```
   |
20 |   buildInputs = [ autoconf automake libtool git ];
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/applications/blockchains/stellar-core.nix:24:15:
```
   |
24 |   patches = [ ./stellar-core-dirty-version.patch ];
   |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

libtool is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/blockchains/stellar-core.nix:20:3:
```
   |
20 |   buildInputs = [ autoconf automake libtool git ];
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

git is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/blockchains/stellar-core.nix:20:3:
```
   |
20 |   buildInputs = [ autoconf automake libtool git ];
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

automake is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/blockchains/stellar-core.nix:20:3:
```
   |
20 |   buildInputs = [ autoconf automake libtool git ];
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/applications/blockchains/stellar-core.nix [hunk 1]
```diff
@@ -3,15 +3,15 @@
...
 let
   pname = "stellar-core";
-  version = "0.5.1";
+  version = "17.0.0";
 
 in stdenv.mkDerivation {
   name = "${pname}-${version}";
 
   src = fetchgit {
     url = "https://github.com/stellar/stellar-core.git";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/stellar-core.nix:12]: Please use fetchFromGitHub.

**Comment by austinbutler** [comment]
Result of `nixpkgs-review pr 123294` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>stellar-core</li>
  </ul>
</details>

File: pkgs/applications/blockchains/stellar-core.nix (hunk 1 above)
Comment (austinbutler) [reply pkgs/applications/blockchains/stellar-core.nix]: Done!

**Comment by jonringer** [comment]
please address the ofborg failure, looks like you have a locally cached sha

**Comment by austinbutler** [comment]
Haven't been able to get my desktop, laptop, or Ofborfg to agree on the hash... re-instated the change to the Makefile from the original PR, although I still don't understand the motivation for that patch.

File: pkgs/applications/blockchains/stellar-core.nix [hunk 2]
```diff
@@ -1,29 +1,25 @@
...
-
-  src = fetchgit {
-    url = "https://github.com/stellar/stellar-core.git";
-    rev = "refs/tags/v${version}";
-    sha256 = "0ldw3qr0sajgam38z2w2iym0214ial6iahbzx3b965cw92n8n88z";
+  src = fetchFromGitHub {
+    owner = "stellar";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1bq5j69zb9bhkqmz7mkz5s47p385szjd9pvm3pp3h898qc5m6am8";
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/stellar-core.nix:12]:
```suggestion
    sha256 = "09vpznc0iw6y8lg0h39yfrjacapvs47wl382vbfg1xrfi1hks30l";
```

Comment (veprbl) [reply pkgs/applications/blockchains/stellar-core.nix:12]: Get `1d1r5q1pac8bxffaz075980dfi4f8sh8vj8gcc1cz4icvsby1x50` here. Could be an instance of #8567

Comment (austinbutler) [reply pkgs/applications/blockchains/stellar-core.nix:12]: I'm glad it's not just me somehow! Thank you for that link, I will try out some workarounds.

File: pkgs/applications/blockchains/stellar-core.nix [hunk 3]
```diff
@@ -1,31 +1,26 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1ngl8yjqb8xzhdwzlxzzxf14q2hgwy2ysb17sn5380rrn0jswin1";
     fetchSubmodules = true;
-    leaveDotGit = true;
   };
 
-  nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ autoconf automake libtool git ];
+  nativeBuildInputs = [ automake autoconf git libtool pkg-config ];
```
Comment (veprbl) [inline pkgs/applications/blockchains/stellar-core.nix:16]:
```suggestion
  nativeBuildInputs = [ autoreconfHook pkg-config ];
```
and drop `preConfigure`

Comment (austinbutler) [reply pkgs/applications/blockchains/stellar-core.nix:16]: Fails to build:

```
automake: error: cannot open < lib/medida.mk: No such file or directory
autoreconf: error: automake failed with exit status: 1
builder for '/nix/store/gmd8jwvgdp8x2fh554978960gmr6mdxi-stellar-core-17.0.0.drv' failed with exit code 1
error: build of '/nix/store/gmd8jwvgdp8x2fh554978960gmr6mdxi-stellar-core-17.0.0.drv' failed
```

**Comment by austinbutler** [comment]
Result of `nixpkgs-review pr 123294` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>stellar-core</li>
    <li>tests.trivial</li>
  </ul>
</details>

**Comment by austinbutler** [comment]
At this point everything seems fine (it builds with a reliable hash, it runs), except the version `stellar-core` thinks it is at is wrong, probably due to the removal of `.git` (which is where the hash failures came from). It tells you it's at version 0.1. Looking for a workaround now.

**Comment by austinbutler** [comment]
I've worked around the version issue as well, so I think this is ready.

**Pull request merged**